### PR TITLE
Implement buzzer methods with AudioContext API

### PR DESCRIPTION
**Pull request by a2**
This PR adds buzzer support for the Sensor Watch simulator. I used the `apps/buzzer-test` app for testing.

### CODE REVIEW

File: apps/buzzer-test/app.c [hunk 1]
```diff
@@ -115,7 +115,7 @@ bool app_loop(void) {
             900,
         };
         application_state.play = false;
-        for(size_t i = 0; i < sizeof(rains); i++) {
+        for(size_t i = 0, count = sizeof(rains) / sizeof(rains[0]); i < count; i++) {
```
Comment (a2) [inline apps/buzzer-test/app.c:118]: I think this worked before because the ARM C compiler treated each note enum as an `int8_t` due to its size. From my testing, the Emscripten C compiler behaves differently.

File: rules.mk [hunk 2]
```diff
@@ -13,6 +13,7 @@ endif
 $(BUILD)/$(BIN).html: $(OBJS)
 	@echo HTML $@
 	@$(CC) $(LDFLAGS) $(OBJS) $(LIBS) -o $@ \
+		-s ASYNCIFY=1 \
```
Comment (a2) [inline rules.mk:16]: This adds support for the `emscripten_sleep()` function.

File: watch-library/simulator/watch/watch_deepsleep.c [hunk 3]
```diff
@@ -35,6 +35,7 @@ static uint32_t watch_backup_data[8];
 
 void watch_register_extwake_callback(uint8_t pin, ext_irq_cb_t callback, bool level) {
     if (pin == BTN_ALARM) {
+        watch_enable_external_interrupts();
```
Comment (a2) [inline watch-library/simulator/watch/watch_deepsleep.c:38]: If a watch app explicitly uses `watch_register_extwake_callback()` to wake with the Alarm button, we need to enable interrupts (a simulator implementation detail).

Comment (joeycastillo) [reply watch-library/simulator/watch/watch_deepsleep.c:38]: right because on the hardware, that interrupt is part of the RTC's responsibilities, makes sense!

**Comment by joeycastillo** [comment]
So with this one, I am afraid I'm getting an exception when I press the Mode button. Seems unrelated to the next watch face; it also happens when the setting or voltage face is next in the carousel. Tested in Safari and Firefox.

Screen Shot 2022-01-25 at 6 40 25 PM

**Comment by joeycastillo** [comment]
First half of that stack trace:

Screen Shot 2022-01-25 at 6 43 22 PM

**Comment by a2** [comment]
Spent some interesting debugging time on this one. I _think_ that there is a race condition (which, of course it is) due to [reentrancy](https://emscripten.org/docs/porting/asyncify.html#reentrancy). (Ehe event handlers are still receiving and processing events while the `emscripten_sleep()` call is waiting.) Tomorrow I'll try refactoring the event handling code to be in "pure" (inline) JavaScript again so I can do more to prevent this condition.

**Comment by a2** [comment]
tl;dr: I tested `movement` and `buzzer-test` and both seem to be working even when I spam the A/L/M keys or lick on the SVG buttons. I also updated some of the interval/timeout/rAF IDs to check against `-1` instead of `0` because `0` is a valid "first" ID.

I think the original issue was the simulator code has a few "entrances": the main loop, which is driven by `requestAnimationFrame()`, and the event handlers. If one of those is triggered during `emscripten_sleep()` (the underlying delay mechanism) then (I think) the `Asyncify` state corrupts and the code is not able to "unwind" correctly after `emscripten_sleep()` completes.

**Comment by joeycastillo** [comment]
Tested both the buzzer test app and Movement in Firefox and on hardware, both seem to work great! Merging…
